### PR TITLE
feat(manager/regex): handle version and digest in separate matchStrings

### PR DESCRIPTION
**Pull request by amezin**
## Changes

When using regex manager with `"matchStringsStrategy": "combination"`, and `currentDigest` and `currentValue` groups in separate `matchString`s, the returned `replaceString` will now include both. Previously, `replaceString` included only one of the groups/matches (the last one), which then caused update failure.

## Context

https://github.com/renovatebot/renovate/discussions/27577

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by amezin** [comment]
Test repository: https://github.com/amezin/renovate-regex-multiline

Successful branch update with this PR applied: https://github.com/amezin/renovate-regex-multiline/commit/b6ca4bbcfee478465379ca51d84aa47924829919 (but then Renovate hung while creating a PR - I guess it's an unrelated issue with http throttling)

**Review by rarkins — COMMENTED** [review]
I think this is a feature, not a fix, including because of the complexity of the change. This will need to wait for a careful review by @secustor when he's back from vacation in ~2 weeks

**Comment by amezin** [comment]
It likely won't work with `autoReplaceStringTemplate`... And it's not clear how replacing multiple strings should work with `autoReplaceStringTemplate`.

**Comment by secustor** [comment]
You could try to find the first occurence of the matchStrings for `currentValue` and `currentDigest` and then use the content in between ( including the actual full matches ) to calculate the `autoreplaceStringTemplate`.

That should work perfectly fine as `combination` only supports one occurrence per file.

**Comment by amezin** [comment]
>  to calculate the autoreplaceStringTemplate

If I'm not mistaken, it worked as is if `autoreplaceStringTemplate` wasn't set in config. What I'm worried about is surprising behavior when the user sets `autoreplaceStringTemplate` themselves.

Or did you mean modifying the user-provided `autoreplaceStringTemplate` somehow?

**Comment by secustor** [comment]
I'm reffering to the calculated one here: https://github.com/renovatebot/renovate/pull/27754/files#diff-115f4ca8ea17ed0c78cc1c54a3d5d101bd2e31b9d2c53d0ad1c8c95bb90c2f0dR65

**Comment by amezin** [comment]
What the code does currently is taking the last match for `currentValue` and the last match for `digest`. And then do what you described - but it's for the last matches, not the first ones.

I'm taking the last matches because previously `combination` worked exactly like that for cases when there's only `currentValue` or only `currentDigest` or both are in the same match. I.e. if I take first matches the behavior will change for the cases where `combination` already worked.

**Comment by amezin** [comment]
Is `autoreplaceStringTemplate` required in addition to `replaceString`? Should it be used instead of `replaceString`?

File: lib/modules/manager/custom/regex/utils.ts [hunk 1]
```diff
@@ -85,9 +85,9 @@ export function createDependency(
 export function regexMatchAll(
   regex: RegExp,
   content: string,
-): RegExpMatchArray[] {
-  const matches: RegExpMatchArray[] = [];
-  let matchResult: RegExpMatchArray | null;
+): RegExpExecArray[] {
```
Comment (viceice) [inline lib/modules/manager/custom/regex/utils.ts:88]: why this change?

Comment (amezin) [reply lib/modules/manager/custom/regex/utils.ts:88]: In `RegExpMatchArray` `index` is optional. And this change needs indices.

Because `RegExp.exec` returns `RegExpExecArray | null`, I thought it's ok to change the return type to `RegExpExecArray[]`

**Comment by amezin** [comment]
It seems that there's been some misunderstanding (primarily on my side).

I just never used `autoReplaceStringTemplate` myself. And also I think I've seen some tricky use of it together with `combination` strategy in the docs. But I can't find it anymore (maybe it actually was never there, and my memory is simply wrong).

**Comment by rarkins** [comment]
Closing due to inactivity + conflicts. @amezin we are happy to discuss this in a Discussion first before reopening, also please feel free to email me for a Slack channel invite if you like
